File: android/src/main/kotlin/com/gazepoint/sdk/GazeTracker.kt
package com.gazepoint.sdk

import android.content.Context
import android.graphics.PointF
import android.util.Log
import com.gazepoint.sdk.math.Vector3
import com.google.mlkit.vision.face.Face
import com.google.mlkit.vision.face.FaceLandmark
import kotlin.math.abs
import kotlin.math.sqrt

/**
 * Enhanced Gaze Tracker with improved accuracy and calibration support.
 *
 * Features:
 * - Kalman filtering for smooth tracking
 * - Multi-point calibration support
 * - Head pose compensation
 * - Adaptive smoothing based on movement speed
 * - Eye aspect ratio for blink detection
 */
class GazeTracker(private val context: Context) {

    companion object {
        private const val TAG = "GazeTracker"
        private const val SMOOTHING_FACTOR = 0.3f
        private const val BLINK_EAR_THRESHOLD = 0.2f
        private const val VELOCITY_THRESHOLD = 100f
    }

    private var lastGazePoint: PointF? = null
    private var calibrationData: CalibrationData? = null
    private var isCalibrated = false
    private var blinkDetected = false
    private val kalmanFilter = KalmanFilter()

    data class CalibrationData(
        val offsetX: Float = 0f,
        val offsetY: Float = 0f,
        val scaleX: Float = 1f,
        val scaleY: Float = 1f,
        val rotationCompensation: Float = 0f
    )

    data class GazeResult(
        val gazePoint: PointF,
        val confidence: Float,
        val isBlinking: Boolean,
        val headPose: HeadPose
    )

    data class HeadPose(
        val pitch: Float,
        val yaw: Float,
        val roll: Float
    )

    /**
     * Calculate gaze point from a detected ML Kit [Face].
     */
    fun calculateGazePoint(face: Face): GazeResult? {
        val leftEyeLandmark = face.getLandmark(FaceLandmark.LEFT_EYE)
        val rightEyeLandmark = face.getLandmark(FaceLandmark.RIGHT_EYE)

        if (leftEyeLandmark == null || rightEyeLandmark == null) {
            Log.w(TAG, "Eye landmarks not detected")
            return null
        }

        val leftEyePosition = leftEyeLandmark.position
        val rightEyePosition = rightEyeLandmark.position

        blinkDetected = detectBlink(face)
        val headPose = calculateHeadPose(face)
        val gazeVector = calculateGazeVector(leftEyePosition, rightEyePosition, headPose)

        val calibratedVector = if (isCalibrated && calibrationData != null) {
            applyCalibration(gazeVector, calibrationData!!)
        } else {
            gazeVector
        }

        val screenPoint = mapGazeVectorToScreenCoordinates(calibratedVector, headPose)
        val filteredPoint = kalmanFilter.update(screenPoint)
        val smoothedPoint = applyAdaptiveSmoothing(filteredPoint)
        val confidence = calculateConfidence(face, leftEyePosition, rightEyePosition)

        lastGazePoint = smoothedPoint

        return GazeResult(
            gazePoint = smoothedPoint,
            confidence = confidence,
            isBlinking = blinkDetected,
            headPose = headPose
        )
    }

    private fun calculateGazeVector(
        leftEyePosition: PointF,
        rightEyePosition: PointF,
        headPose: HeadPose
    ): Vector3 {
        val baseVector = Vector3(
            rightEyePosition.x - leftEyePosition.x,
            rightEyePosition.y - leftEyePosition.y,
            0f
        )

        val compensatedX = baseVector.x + headPose.yaw * 0.5f
        val compensatedY = baseVector.y + headPose.pitch * 0.5f
        val gazeVector = Vector3(compensatedX, compensatedY, 0f)

        val magnitude = sqrt(gazeVector.x * gazeVector.x + gazeVector.y * gazeVector.y)
        if (magnitude > 0) {
            return Vector3(gazeVector.x / magnitude, gazeVector.y / magnitude, 0f)
        }

        return gazeVector
    }

    private fun calculateHeadPose(face: Face): HeadPose {
        return HeadPose(
            pitch = face.headEulerAngleX,
            yaw = face.headEulerAngleY,
            roll = face.headEulerAngleZ
        )
    }

    private fun detectBlink(face: Face): Boolean {
        val leftEyeOpen = face.leftEyeOpenProbability ?: 1f
        val rightEyeOpen = face.rightEyeOpenProbability ?: 1f
        val avgEyeOpen = (leftEyeOpen + rightEyeOpen) / 2f
        return avgEyeOpen < BLINK_EAR_THRESHOLD
    }

    private fun mapGazeVectorToScreenCoordinates(
        gazeVector: Vector3,
        headPose: HeadPose
    ): PointF {
        val displayMetrics = context.resources.displayMetrics
        val screenWidth = displayMetrics.widthPixels.toFloat()
        val screenHeight = displayMetrics.heightPixels.toFloat()

        val yawFactor = 1f + (abs(headPose.yaw) / 30f) * 0.2f
        val pitchFactor = 1f + (abs(headPose.pitch) / 30f) * 0.2f

        val screenX = (screenWidth / 2f) + (gazeVector.x * (screenWidth / 2f) * yawFactor)
        val screenY = (screenHeight / 2f) - (gazeVector.y * (screenHeight / 2f) * pitchFactor)

        return PointF(
            screenX.coerceIn(0f, screenWidth),
            screenY.coerceIn(0f, screenHeight)
        )
    }

    private fun applyCalibration(gazeVector: Vector3, calibration: CalibrationData): Vector3 {
        return Vector3(
            gazeVector.x * calibration.scaleX + calibration.offsetX,
            gazeVector.y * calibration.scaleY + calibration.offsetY,
            gazeVector.z
        )
    }

    private fun applyAdaptiveSmoothing(currentPoint: PointF): PointF {
        val lastPoint = lastGazePoint ?: return currentPoint

        val dx = currentPoint.x - lastPoint.x
        val dy = currentPoint.y - lastPoint.y
        val velocity = sqrt(dx * dx + dy * dy)

        val adaptiveFactor = if (velocity > VELOCITY_THRESHOLD) {
            SMOOTHING_FACTOR * 0.5f
        } else {
            SMOOTHING_FACTOR
        }

        return PointF(
            lastPoint.x + (currentPoint.x - lastPoint.x) * adaptiveFactor,
            lastPoint.y + (currentPoint.y - lastPoint.y) * adaptiveFactor
        )
    }

    private fun calculateConfidence(
        face: Face,
        leftEyePosition: PointF,
        rightEyePosition: PointF
    ): Float {
        var confidence = 1.0f

        confidence *= (face.trackingId ?: 0).let { if (it > 0) 1.0f else 0.8f }

        lastGazePoint?.let { last ->
            val eyeMidX = (leftEyePosition.x + rightEyePosition.x) / 2f
            val eyeMidY = (leftEyePosition.y + rightEyePosition.y) / 2f
            val stability = 1f - (abs(last.x - eyeMidX) + abs(last.y - eyeMidY)) / 1000f
            confidence *= stability.coerceIn(0.5f, 1.0f)
        }

        if (blinkDetected) {
            confidence *= 0.3f
        }

        return confidence.coerceIn(0f, 1f)
    }

    /**
     * Calibrate with known screen points: pairs of (expected, actual).
     * Requires at least 3 points.
     */
    fun calibrate(calibrationPoints: List<Pair<PointF, PointF>>) {
        if (calibrationPoints.size < 3) {
            Log.w(TAG, "Need at least 3 calibration points")
            return
        }

        var sumOffsetX = 0f
        var sumOffsetY = 0f
        var sumScaleX = 0f
        var sumScaleY = 0f

        calibrationPoints.forEach { (expected, actual) ->
            sumOffsetX += expected.x - actual.x
            sumOffsetY += expected.y - actual.y
            if (actual.x != 0f) sumScaleX += expected.x / actual.x
            if (actual.y != 0f) sumScaleY += expected.y / actual.y
        }

        val count = calibrationPoints.size
        calibrationData = CalibrationData(
            offsetX = sumOffsetX / count,
            offsetY = sumOffsetY / count,
            scaleX = sumScaleX / count,
            scaleY = sumScaleY / count
        )

        isCalibrated = true
        Log.i(TAG, "Calibration completed: $calibrationData")
    }

    fun resetCalibration() {
        calibrationData = null
        isCalibrated = false
        lastGazePoint = null
        kalmanFilter.reset()
    }

    private class KalmanFilter {
        private var estimateX = 0f
        private var estimateY = 0f
        private var errorCovarianceX = 1f
        private var errorCovarianceY = 1f

        private val processNoise = 0.01f
        private val measurementNoise = 0.1f

        fun update(measurement: PointF): PointF {
            val predictedErrorCovX = errorCovarianceX + processNoise
            val predictedErrorCovY = errorCovarianceY + processNoise

            val kalmanGainX = predictedErrorCovX / (predictedErrorCovX + measurementNoise)
            val kalmanGainY = predictedErrorCovY / (predictedErrorCovY + measurementNoise)

            estimateX += kalmanGainX * (measurement.x - estimateX)
            estimateY += kalmanGainY * (measurement.y - estimateY)

            errorCovarianceX = (1 - kalmanGainX) * predictedErrorCovX
            errorCovarianceY = (1 - kalmanGainY) * predictedErrorCovY

            return PointF(estimateX, estimateY)
        }

        fun reset() {
            estimateX = 0f
            estimateY = 0f
            errorCovarianceX = 1f
            errorCovarianceY = 1f
        }
    }
}

File: macos/gazepoint_sdk/Sources/gazepoint_sdk/GazePointSDK/GazeTracker.swift
import AppKit
import Foundation
import Vision

/// Gaze tracker for macOS using Vision face landmarks.
@available(macOS 12.0, *)
public class GazeTracker {

    public struct GazeResult {
        public let gazePoint: CGPoint
        public let confidence: Float
        public let isBlinking: Bool
        public let headPose: HeadPose
        public let timestamp: TimeInterval

        public init(
            gazePoint: CGPoint,
            confidence: Float,
            isBlinking: Bool,
            headPose: HeadPose,
            timestamp: TimeInterval
        ) {
            self.gazePoint = gazePoint
            self.confidence = confidence
            self.isBlinking = isBlinking
            self.headPose = headPose
            self.timestamp = timestamp
        }
    }

    public struct HeadPose {
        public let pitch: Float
        public let yaw: Float
        public let roll: Float

        public init(pitch: Float, yaw: Float, roll: Float) {
            self.pitch = pitch
            self.yaw = yaw
            self.roll = roll
        }
    }

    public struct CalibrationData: Codable {
        public var offsetX: Float
        public var offsetY: Float
        public var scaleX: Float
        public var scaleY: Float
        public var rotationCompensation: Float

        public init(
            offsetX: Float = 0,
            offsetY: Float = 0,
            scaleX: Float = 1,
            scaleY: Float = 1,
            rotationCompensation: Float = 0
        ) {
            self.offsetX = offsetX
            self.offsetY = offsetY
            self.scaleX = scaleX
            self.scaleY = scaleY
            self.rotationCompensation = rotationCompensation
        }
    }

    private let smoothingFactor: Float = 0.3
    private let blinkThreshold: Float = 0.3
    private let velocityThreshold: Float = 100.0

    private var lastGazePoint: CGPoint?
    private var calibrationData: CalibrationData?
    private var isCalibrated: Bool = false
    private var kalmanFilter: KalmanFilter
    private let performanceMonitor: PerformanceMonitor

    private lazy var faceDetectionRequest: VNDetectFaceLandmarksRequest = {
        let request = VNDetectFaceLandmarksRequest()
        request.revision = VNDetectFaceLandmarksRequestRevision3
        return request
    }()

    public init() {
        self.kalmanFilter = KalmanFilter()
        self.performanceMonitor = PerformanceMonitor()
    }

    public func calculateGazePoint(
        from pixelBuffer: CVPixelBuffer,
        orientation: CGImagePropertyOrientation = .up
    ) -> GazeResult? {
        let startTime = performanceMonitor.startFrame()
        defer {
            performanceMonitor.endFrame(startTime: startTime)
        }

        let handler = VNImageRequestHandler(
            cvPixelBuffer: pixelBuffer,
            orientation: orientation,
            options: [:]
        )

        do {
            try handler.perform([faceDetectionRequest])
            guard let faceObservation = faceDetectionRequest.results?.first else {
                return nil
            }
            return processGaze(from: faceObservation)
        } catch {
            return nil
        }
    }

    public func calibrate(calibrationPoints: [(expected: CGPoint, actual: CGPoint)]) {
        guard calibrationPoints.count >= 3 else { return }

        var sumOffsetX: Float = 0
        var sumOffsetY: Float = 0
        var sumScaleX: Float = 0
        var sumScaleY: Float = 0

        for (expected, actual) in calibrationPoints {
            sumOffsetX += Float(expected.x - actual.x)
            sumOffsetY += Float(expected.y - actual.y)
            if actual.x != 0 {
                sumScaleX += Float(expected.x / actual.x)
            }
            if actual.y != 0 {
                sumScaleY += Float(expected.y / actual.y)
            }
        }

        let count = Float(calibrationPoints.count)
        calibrationData = CalibrationData(
            offsetX: sumOffsetX / count,
            offsetY: sumOffsetY / count,
            scaleX: sumScaleX / count,
            scaleY: sumScaleY / count
        )
        isCalibrated = true
    }

    public func resetCalibration() {
        calibrationData = nil
        isCalibrated = false
        lastGazePoint = nil
        kalmanFilter.reset()
    }

    public func getPerformanceMetrics() -> PerformanceMonitor.PerformanceMetrics {
        return performanceMonitor.getMetrics()
    }

    private func processGaze(from faceObservation: VNFaceObservation) -> GazeResult? {
        guard let landmarks = faceObservation.landmarks,
              let leftEye = landmarks.leftEye,
              let rightEye = landmarks.rightEye else {
            return nil
        }

        let isBlinking = detectBlink(leftEye: leftEye, rightEye: rightEye)
        let headPose = calculateHeadPose(from: faceObservation)
        let gazeVector = calculateGazeVector(
            leftEye: leftEye,
            rightEye: rightEye,
            headPose: headPose
        )
        let calibratedVector = applyCalibration(to: gazeVector)
        let screenSize = NSScreen.main?.frame.size ?? CGSize(width: 1440, height: 900)
        let screenPoint = mapGazeVectorToScreen(
            gazeVector: calibratedVector,
            headPose: headPose,
            screenSize: screenSize
        )
        let filteredPoint = kalmanFilter.update(measurement: screenPoint)
        let smoothedPoint = applyAdaptiveSmoothing(currentPoint: filteredPoint)
        let confidence = calculateConfidence(
            faceObservation: faceObservation,
            isBlinking: isBlinking
        )
        lastGazePoint = smoothedPoint

        return GazeResult(
            gazePoint: smoothedPoint,
            confidence: confidence,
            isBlinking: isBlinking,
            headPose: headPose,
            timestamp: Date().timeIntervalSince1970
        )
    }

    private func calculateGazeVector(
        leftEye: VNFaceLandmarkRegion2D,
        rightEye: VNFaceLandmarkRegion2D,
        headPose: HeadPose
    ) -> CGPoint {
        let leftPoints = leftEye.normalizedPoints
        let rightPoints = rightEye.normalizedPoints
        guard !leftPoints.isEmpty, !rightPoints.isEmpty else { return .zero }

        let leftCenter = averagePoint(points: leftPoints)
        let rightCenter = averagePoint(points: rightPoints)

        var gazeX = rightCenter.x - leftCenter.x
        var gazeY = rightCenter.y - leftCenter.y
        gazeX += CGFloat(headPose.yaw) * 0.005
        gazeY += CGFloat(headPose.pitch) * 0.005

        let magnitude = sqrt(gazeX * gazeX + gazeY * gazeY)
        if magnitude > 0 {
            gazeX /= magnitude
            gazeY /= magnitude
        }
        return CGPoint(x: gazeX, y: gazeY)
    }

    private func calculateHeadPose(from faceObservation: VNFaceObservation) -> HeadPose {
        HeadPose(
            pitch: faceObservation.pitch?.floatValue ?? 0,
            yaw: faceObservation.yaw?.floatValue ?? 0,
            roll: faceObservation.roll?.floatValue ?? 0
        )
    }

    private func detectBlink(
        leftEye: VNFaceLandmarkRegion2D,
        rightEye: VNFaceLandmarkRegion2D
    ) -> Bool {
        let avgEAR = (calculateEyeAspectRatio(eye: leftEye) + calculateEyeAspectRatio(eye: rightEye)) / 2
        return avgEAR < blinkThreshold
    }

    private func calculateEyeAspectRatio(eye: VNFaceLandmarkRegion2D) -> Float {
        let points = eye.normalizedPoints
        guard points.count >= 6 else { return 1.0 }
        let vertical1 = distance(from: points[1], to: points[5])
        let vertical2 = distance(from: points[2], to: points[4])
        let horizontal = distance(from: points[0], to: points[3])
        if horizontal == 0 { return 1.0 }
        return Float((vertical1 + vertical2) / (2.0 * horizontal))
    }

    private func mapGazeVectorToScreen(
        gazeVector: CGPoint,
        headPose: HeadPose,
        screenSize: CGSize
    ) -> CGPoint {
        let yawFactor: CGFloat = 1.0 + (CGFloat(abs(headPose.yaw)) / 30.0) * 0.2
        let pitchFactor: CGFloat = 1.0 + (CGFloat(abs(headPose.pitch)) / 30.0) * 0.2
        var screenX = (screenSize.width / 2) + (gazeVector.x * (screenSize.width / 2) * yawFactor)
        var screenY = (screenSize.height / 2) - (gazeVector.y * (screenSize.height / 2) * pitchFactor)
        screenX = max(0, min(screenSize.width, screenX))
        screenY = max(0, min(screenSize.height, screenY))
        return CGPoint(x: screenX, y: screenY)
    }

    private func applyCalibration(to gazeVector: CGPoint) -> CGPoint {
        guard isCalibrated, let calibration = calibrationData else {
            return gazeVector
        }
        return CGPoint(
            x: CGFloat(gazeVector.x * CGFloat(calibration.scaleX) + CGFloat(calibration.offsetX)),
            y: CGFloat(gazeVector.y * CGFloat(calibration.scaleY) + CGFloat(calibration.offsetY))
        )
    }

    private func applyAdaptiveSmoothing(currentPoint: CGPoint) -> CGPoint {
        guard let lastPoint = lastGazePoint else { return currentPoint }
        let dx = currentPoint.x - lastPoint.x
        let dy = currentPoint.y - lastPoint.y
        let velocity = sqrt(dx * dx + dy * dy)
        let adaptiveFactor: CGFloat = velocity > CGFloat(velocityThreshold)
            ? CGFloat(smoothingFactor) * 0.5
            : CGFloat(smoothingFactor)
        return CGPoint(
            x: lastPoint.x + (currentPoint.x - lastPoint.x) * adaptiveFactor,
            y: lastPoint.y + (currentPoint.y - lastPoint.y) * adaptiveFactor
        )
    }

    private func calculateConfidence(
        faceObservation: VNFaceObservation,
        isBlinking: Bool
    ) -> Float {
        var confidence = faceObservation.confidence
        if isBlinking {
            confidence *= 0.3
        }
        let faceArea = faceObservation.boundingBox.width * faceObservation.boundingBox.height
        if faceArea < 0.05 {
            confidence *= 0.7
        }
        return max(0, min(1, confidence))
    }

    private func averagePoint(points: [CGPoint]) -> CGPoint {
        guard !points.isEmpty else { return .zero }
        var sumX: CGFloat = 0
        var sumY: CGFloat = 0
        for point in points {
            sumX += point.x
            sumY += point.y
        }
        return CGPoint(x: sumX / CGFloat(points.count), y: sumY / CGFloat(points.count))
    }

    private func distance(from p1: CGPoint, to p2: CGPoint) -> CGFloat {
        let dx = p2.x - p1.x
        let dy = p2.y - p1.y
        return sqrt(dx * dx + dy * dy)
    }
}
